cpToPod failed after 30 attempts (related to permission issue)

I have been trying to implement https://github.com/actions/runner-container-hooks/pull/244 by using version `0.8.0` of `runner-container-hooks`.

However I have been struggling so far to get it to work.
This is the template I have been using.

```yaml
template:
  spec:
    securityContext:
        fsGroup: 1000
    serviceAccountName: xxx-redacted
    containers:
    - name: runner
      image: my-company.registry.com/arc-linux-runner:2.328.0-3dacb977f4376955e068080cf4e8a42efe84e83d
      command: ["/home/runner/run.sh"]
      resources:
        limits:
          ephemeral-storage: 2Gi
          memory: 512Mi
        requests:
          cpu: 50m
          ephemeral-storage: 2Gi
          memory: 256Mi
      env:
        - name: ACTIONS_RUNNER_USE_KUBE_SCHEDULER
          value: "true"
        - name: ACTIONS_RUNNER_PREPARE_JOB_TIMEOUT_SECONDS
          value: "1800"
        - name: ACTIONS_RUNNER_CONTAINER_HOOKS
          value: /home/runner/k8s/index.js
        - name: ACTIONS_RUNNER_POD_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
        - name: ACTIONS_RUNNER_REQUIRE_JOB_CONTAINER
          value: "true"
        - name: ACTIONS_RUNNER_CONTAINER_HOOK_TEMPLATE
          value: /home/runner/hook-extension.yaml
      volumeMounts:
        - name: hook-extension
          mountPath: /home/runner/hook-extension.yaml
          subPath: content
    nodeSelector:
      kubernetes.io/arch: amd64
    volumes:
      - name: hook-extension
        configMap:
          name: hook-extension-{{ gha_runner_scale_set__name }}
```

The `Initialize Containers` step always fails with the following message:
```
Run '/home/runner/k8s/index.js'
(node:66) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
Error: Error: cpToPod failed after 30 attempts: {}
Error: Process completed with exit code 1.
Error: Executing the custom container implementation failed. Please contact your self hosted runner administrator
```

By enabling debug mode it becomes clear that the issue is related to permissions.
```
##[debug]cpToPod: Attempt 1 failed: Error: Error from cpFromPod - details: 
##[debug] tar: _PipelineMapping: Cannot mkdir: Permission denied
##[debug]tar: _temp: Cannot mkdir: Permission denied
##[debug]tar: _tool: Cannot mkdir: Permission denied
##[debug]tar: tm-pnt-dummy: Cannot utime: Operation not permitted
##[debug]tar: _PipelineMapping: Cannot mkdir: Permission denied
##[debug]tar: _PipelineMapping/TrendMiner: Cannot mkdir: No such file or directory
##[debug]tar: _temp: Cannot mkdir: Permission denied
##[debug]tar: _temp/_github_home: Cannot mkdir: No such file or directory
##[debug]tar: _temp: Cannot mkdir: Permission denied
##[debug]tar: _temp/_github_workflow: Cannot mkdir: No such file or directory
##[debug]tar: _temp: Cannot mkdir: Permission denied
##[debug]tar: _temp/_runner_hook_responses: Cannot mkdir: No such file or directory
##[debug]tar: tm-pnt-dummy/tm-pnt-dummy: Cannot utime: Operation not permitted
##[debug]tar: _PipelineMapping/TrendMiner: Cannot mkdir: No such file or directory
##[debug]tar: _PipelineMapping/TrendMiner/tm-pnt-dummy: Cannot mkdir: No such file or directory
##[debug]tar: _temp/_runner_hook_responses: Cannot mkdir: No such file or directory
##[debug]tar: _temp/_runner_hook_responses/aa549ea6-b227-441e-ab3f-7ca69acc9602.json: Cannot open: No such file or directory
##[debug]tar: _PipelineMapping/TrendMiner: Cannot mkdir: No such file or directory
##[debug]tar: _PipelineMapping/TrendMiner/tm-pnt-dummy/PipelineFolder.json: Cannot open: No such file or directory
##[debug]tar: .: Cannot utime: Operation not permitted
##[debug]tar: Exiting with failure status due to previous errors
```

The workflow that I am using is the following.

```yaml
name: Troubleshoot autoscaling
on:
  workflow_dispatch:
jobs:
  job1:
    runs-on: default-staging
    container: company.registry.com/tm-gha-base:v1
    steps:
      - name: Use a big runner to test autoscaler
        run: sleep 600
```

Our `tm-gha-base` container is an ubuntu based container which runs as `uid` 1000 and thus not as root.
As you can see in the scaleset configuration, we are also defining a `securityContext` with `fsGroup` set to `1000`.

Is there any support to use non-root containers, like we are using now?

--- a/package.json
+++ b/package.json
@@ -12,6 +12,7 @@
     "format": "prettier --write '**/*.ts'",
     "format-check": "prettier --check '**/*.ts'",
     "lint": "eslint packages/**/*.ts",
+    "lint:fix": "eslint packages/**/*.ts --fix",
     "build-all": "npm run build --prefix packages/hooklib && npm run build --prefix packages/k8s && npm run build --prefix packages/docker"
   },
   "repository": {

--- a/packages/k8s/src/k8s/index.ts
+++ b/packages/k8s/src/k8s/index.ts
@@ -102,7 +102,9 @@ export async function createJobPod(
       image:
         process.env.ACTIONS_RUNNER_IMAGE ||
         'ghcr.io/actions/actions-runner:latest',
-      command: ['sh', '-c', 
+      command: [
+        'sh',
+        '-c',
         `mkdir -p /mnt/externals && \\
          mkdir -p /mnt/work && \\
          mkdir -p /mnt/github && \\
@@ -375,10 +377,12 @@ export async function execCpToPod(
       const exec = new k8s.Exec(kc)
       // Use tar to extract with --no-same-owner to avoid ownership issues.
       // Then use find to fix permissions. The -m flag helps but we also need to fix permissions after.
-      const command = ['sh', '-c', 
+      const command = [
+        'sh',
+        '-c',
         `tar xf - --no-same-owner -C ${shlex.quote(containerPath)} 2>/dev/null; ` +
-        `find ${shlex.quote(containerPath)} -type f -exec chmod u+rw {} \\; 2>/dev/null; ` +
-        `find ${shlex.quote(containerPath)} -type d -exec chmod u+rwx {} \\; 2>/dev/null`
+          `find ${shlex.quote(containerPath)} -type f -exec chmod u+rw {} \\; 2>/dev/null; ` +
+          `find ${shlex.quote(containerPath)} -type d -exec chmod u+rwx {} \\; 2>/dev/null`
       ]
       const readStream = tar.pack(runnerPath)
       const errStream = new WritableStreamBuffer()

--- a/packages/k8s/src/k8s/index.ts
+++ b/packages/k8s/src/k8s/index.ts
@@ -20,8 +20,10 @@ import {
   listDirAllCommand,
   sleep,
   EXTERNALS_VOLUME_NAME,
-  GITHUB_VOLUME_NAME
+  GITHUB_VOLUME_NAME,
+  WORK_VOLUME
 } from './utils'
+import * as shlex from 'shlex'
 
 const kc = new k8s.KubeConfig()
 
@@ -91,13 +93,21 @@ export async function createJobPod(
 
   appPod.spec = new k8s.V1PodSpec()
   appPod.spec.containers = containers
+  appPod.spec.securityContext = {
+    fsGroup: 1001
+  }
   appPod.spec.initContainers = [
     {
       name: 'fs-init',
       image:
         process.env.ACTIONS_RUNNER_IMAGE ||
         'ghcr.io/actions/actions-runner:latest',
-      command: ['sh', '-c', 'mv /home/runner/externals/* /mnt/externals'],
+      command: ['sh', '-c', 
+        `mkdir -p /mnt/externals && \\
+         mkdir -p /mnt/work && \\
+         mkdir -p /mnt/github && \\
+         mv /home/runner/externals/* /mnt/externals/`
+      ],
       securityContext: {
         runAsGroup: 1001,
         runAsUser: 1001
@@ -106,6 +116,14 @@ export async function createJobPod(
         {
           name: EXTERNALS_VOLUME_NAME,
           mountPath: '/mnt/externals'
+        },
+        {
+          name: WORK_VOLUME,
+          mountPath: '/mnt/work'
+        },
+        {
+          name: GITHUB_VOLUME_NAME,
+          mountPath: '/mnt/github'
         }
       ]
     }
@@ -120,6 +138,10 @@ export async function createJobPod(
     },
     {
       name: GITHUB_VOLUME_NAME,
+      emptyDir: {}
+    },
+    {
+      name: WORK_VOLUME,
       emptyDir: {}
     }
   ]
@@ -351,7 +373,13 @@ export async function execCpToPod(
   while (true) {
     try {
       const exec = new k8s.Exec(kc)
-      const command = ['tar', 'xf', '-', '-C', containerPath]
+      // Use tar to extract with --no-same-owner to avoid ownership issues.
+      // Then use find to fix permissions. The -m flag helps but we also need to fix permissions after.
+      const command = ['sh', '-c', 
+        `tar xf - --no-same-owner -C ${shlex.quote(containerPath)} 2>/dev/null; ` +
+        `find ${shlex.quote(containerPath)} -type f -exec chmod u+rw {} \\; 2>/dev/null; ` +
+        `find ${shlex.quote(containerPath)} -type d -exec chmod u+rwx {} \\; 2>/dev/null`
+      ]
       const readStream = tar.pack(runnerPath)
       const errStream = new WritableStreamBuffer()
       await new Promise((resolve, reject) => {
@@ -369,7 +397,7 @@ export async function execCpToPod(
               if (errStream.size()) {
                 reject(
                   new Error(
-                    `Error from cpFromPod - details: \n ${errStream.getContentsAsString()}`
+                    `Error from execCpToPod - status: ${status.status}, details: \n ${errStream.getContentsAsString()}`
                   )
                 )
               }

--- a/packages/k8s/src/k8s/index.ts
+++ b/packages/k8s/src/k8s/index.ts
@@ -204,6 +204,10 @@ export async function createContainerStepPod(
     {
       name: GITHUB_VOLUME_NAME,
       emptyDir: {}
+    },
+    {
+      name: WORK_VOLUME,
+      emptyDir: {}
     }
   ]
 

--- a/packages/k8s/src/k8s/utils.ts
+++ b/packages/k8s/src/k8s/utils.ts
@@ -15,11 +15,16 @@ export const ENV_USE_KUBE_SCHEDULER = 'ACTIONS_RUNNER_USE_KUBE_SCHEDULER'
 
 export const EXTERNALS_VOLUME_NAME = 'externals'
 export const GITHUB_VOLUME_NAME = 'github'
+export const WORK_VOLUME = 'work'
 
 export const CONTAINER_VOLUMES: k8s.V1VolumeMount[] = [
   {
     name: EXTERNALS_VOLUME_NAME,
     mountPath: '/__e'
+  },
+  {
+    name: WORK_VOLUME,
+    mountPath: '/__w'
   },
   {
     name: GITHUB_VOLUME_NAME,

--- a/packages/k8s/tests/e2e-test.ts
+++ b/packages/k8s/tests/e2e-test.ts
@@ -26,6 +26,7 @@ describe('e2e', () => {
   afterEach(async () => {
     await testHelper.cleanup()
   })
+
   it('should prepare job, run script step, run container step then cleanup without errors', async () => {
     await expect(
       prepareJob(prepareJobData.args, prepareJobOutputFilePath)

--- a/packages/k8s/tests/prepare-job-test.ts
+++ b/packages/k8s/tests/prepare-job-test.ts
@@ -231,4 +231,16 @@ describe('Prepare job', () => {
       expect(() => content.context.services[0].image).not.toThrow()
     }
   )
+
+  it('should prepare job with container with non-root user', async () => {
+    prepareJobData.args!.container!.image =
+      'ghcr.io/actions/actions-runner:latest' // known to use user 1001
+    await expect(
+      prepareJob(prepareJobData.args!, prepareJobOutputFilePath)
+    ).resolves.not.toThrow()
+
+    const content = JSON.parse(
+      fs.readFileSync(prepareJobOutputFilePath).toString()
+    )
+  })
 })

--- a/packages/k8s/tests/prepare-job-test.ts
+++ b/packages/k8s/tests/prepare-job-test.ts
@@ -236,11 +236,11 @@ describe('Prepare job', () => {
     prepareJobData.args!.container!.image =
       'ghcr.io/actions/actions-runner:latest' // known to use user 1001
     await expect(
-      prepareJob(prepareJobData.args!, prepareJobOutputFilePath)
+      prepareJob(prepareJobData.args, prepareJobOutputFilePath)
     ).resolves.not.toThrow()
 
-    const content = JSON.parse(
+    JSON.parse(
       fs.readFileSync(prepareJobOutputFilePath).toString()
-    )
+    ).resolves.not.toThrow()
   })
 })

--- a/packages/k8s/tests/prepare-job-test.ts
+++ b/packages/k8s/tests/prepare-job-test.ts
@@ -239,8 +239,12 @@ describe('Prepare job', () => {
       prepareJob(prepareJobData.args, prepareJobOutputFilePath)
     ).resolves.not.toThrow()
 
-    JSON.parse(
+    const content = JSON.parse(
       fs.readFileSync(prepareJobOutputFilePath).toString()
-    ).resolves.not.toThrow()
+    )
+    expect(content.state.jobPod).toBeTruthy()
+    expect(content.context.container.image).toBe(
+      'ghcr.io/actions/actions-runner:latest'
+    )
   })
 })

